extractor/image/daemon/image.go

```diff
@@ -0,0 +1,134 @@
+package daemon
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"golang.org/x/xerrors"
+)
+
+var mu sync.Mutex
+
+// image is a wrapper for github.com/google/go-containerregistry/pkg/v1/daemon.Image
+// daemon.Image loads the entire image into the memory at first,
+// but it doesn't need to load it if the information is already in the cache,
+// To avoid entire loading, this wrapper uses ImageInspectWithRaw and checks image ID and layer IDs.
+type image struct {
+	v1.Image
+	opener  opener
+	inspect types.ImageInspect
+}
+
+type opener func() (v1.Image, error)
+
+func Image(ref name.Reference) (v1.Image, func(), error) {
+	c, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("failed to initialize a docker client: %w", err)
+	}
+
+	c.NegotiateAPIVersion(context.Background())
+
+	inspect, _, err := c.ImageInspectWithRaw(context.Background(), ref.Name())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	f, err := ioutil.TempFile("", "fanal-*")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cleanup := func() {
+		_ = os.Remove(f.Name())
+	}
+
+	return &image{
+		opener:  imageOpener(c, ref, f),
+		inspect: inspect,
+	}, cleanup, nil
+}
+
+func imageOpener(c *client.Client, ref name.Reference, f *os.File) opener {
+	return func() (v1.Image, error) {
+		// Store the tarball in local filesystem and return a new reader into the bytes each time we need to access something.
+		rc, err := c.ImageSave(context.Background(), []string{ref.Name()})
+		if err != nil {
+			return nil, xerrors.Errorf("foo: %w", err)
+		}
+		defer rc.Close()
+
+		if _, err = io.Copy(f, rc); err != nil {
+			return nil, xerrors.Errorf("foo: %w", err)
+		}
+		defer f.Close()
+
+		image, err := tarball.ImageFromPath(f.Name(), nil)
+		//image, err := daemon.Image(img.imageRef, daemon.WithClient(img.client), daemon.WithUnbufferedOpener())
+		if err != nil {
+			return nil, xerrors.Errorf("foo: %w", err)
+		}
+
+		return image, nil
+	}
+}
+
+func (img *image) populateImage() (err error) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if img.Image != nil {
+		return nil
+	}
```

Comment:
What does this do?

Reply:
Initializing `img.Image` is a high cost, so we should avoid it as possible as we can. If `img.Image` is not nil, we don't have to initialize it again.
https://github.com/aquasecurity/fanal/pull/96/commits/fdc17a2c8a87d22f309a1f7fbe22c16e4689d303